python/lsst/ip/isr/isrTask.py

```diff
@@ -2080,13 +2108,17 @@ def updateVariance(self, ampExposure, amp, overscanImage=None):
             Amplifier detector data.
         overscanImage : `lsst.afw.image.MaskedImage`, optional.
             Image of overscan, required only for empirical read noise.
-
+        ptcDataset : `lsst.ip.isr.PhotonTransferCurveDataset`, optional
+            PTC dataset containing the gains and read noise.
         See also
         --------
         lsst.ip.isr.isrFunctions.updateVariance
         """
         maskPlanes = [self.config.saturatedMaskName, self.config.suspectMaskName]
-        gain = amp.getGain()
+        if self.config.usePtcGains and ptcDataset is not None:
+            gain = ptcDataset.gain[amp.getName()]
+        else:
+            gain = amp.getGain()
```

Comment:
One of two things needs to happen here: either way, that first if statement needs to be split to check if we're using PTC gains, and then do the `if None` part. What we need to decide here is, if usePtcGains is True and ptcDataset _is_ None, do we log a warning and use the amp gains, or do we raise? I think I quite strongly prefer the raising option, because if you're specifically saying to use them, and can't, that should be a raise (for comparison, you can't say doBias and have that not present, so I think this should behave the same), but if there's an argument for logging a warning and defaulting to amp gains, I'm open to hearing it for sure.

Reply:
I was following the example that was there in which if `if self.config.doEmpiricalReadNoise and overscanImage is not None:` failed then it did not raise, but just defaulted to the amp read noise. So I thought the behavior should be similar.

Reply:
So I'm thinking of doing the first "if" split that you suggest (checking if the user wants to use PTC gains), and if not, then defaulting to amp gains, with a warning.

Reply:
Honestly, I'm leaning quite strongly the other way, and think that the `doEmpiricalReadNoise` thing is doing it incorrectly as well. At the end of the day, a warning is a single line in a (sadly) pretty cluttered output, which is easily missed, and if someone is asking for something specific like this, and it can't happen, then I think that should raise. There are very few examples in our pipelines when we ignore specified options and just continue, and I think that this is a particularly bad place to be doing so, because it'll likely usually make a small enough difference that it might not even be noticed downstream, but if people are using it, they'll be doing so for a reason. @RobertLuptonTheGood and @czwa, what are your thoughts here?

Assuming we do move to a raise here, I think we should ticket changing the way `doEmpiricalReadNoise` works too (or just do it on this ticket?)